[libc++] Make drop_view::begin constant time (#72883)

casavaca opened the pull request:
As pointed out in #72883, the implementation only needs to return the
value of ranges::next and does not need to obtain the value through
ranges::advance, which causes it to have O(n) complexity in the case
of random-access-sized but non-common range.

Fixes #72883

EricWF commented:
Please add tests or benchmarks of some sort.

hawkinsw commented:
Are we missing another optimization here, too? 

For the non-`const` version of `begin()`, it could be possible for `__simple_view<_View>` to be the only one of the three conditions to fail. That would get us to exactly the same place as the requirements for the `const` version of `begin()`, just called on a non-`const` `drop_view`. So, should there be an optimization here, too (as in, obviating the need for caching)?

casavaca commented:
Upon further investigation, I found that the library correctly implement O(1) behavior. The call sequence is
```
ranges::next --> ranges::advance (with sentinel) --> ranges::advance (without sentinel) --> random_access_iterator constexpr branch
```
So, the original behavior is correct, and I'm closing this pull request.

casavaca commented:
Oops, I didn't pay attention to `non-common` range...

hawkinsw commented:
> Oops, I didn't pay attention to `non-common` range...

Which would get us to a position where
1. `ranges::end(__base_)` is of a different type than `ranges::begin(__base_)` 
2. therefore, `ranges::end(__base_)` might not model `random_access_iterator`
3. therefore, it would not meet the constant-time-complexity requirements for `std::ranges::next` (see (4) at https://en.cppreference.com/w/cpp/iterator/ranges/next).

I just wanted to double check the reasoning. Thank you for doing this work!
Will

casavaca committed:
[libc++] optimization on ranges::drop_view::begin (#72883)

The optimization is specifically for non common range,
as pointed out from the github issue:
> ... the implementation only needs to return the value of ranges::next
  and does not need to obtain the value through ranges::advance, which will have
  O(n) complexity in the case of random-access-sized but non-common range.

casavaca commented:
OK. Hopefully I did it right this time. First time contributor here.

- [X] git clang-format
- [X] 2 changes at drop_view.h are covered by 2 newly added test cases respectively.

huixie90 commented on libcxx/include/__ranges/drop_view.h:
```diff
@@ -90,6 +90,10 @@ namespace ranges {
       requires (!(__simple_view<_View> &&
                   random_access_range<const _View> && sized_range<const _View>))
     {
+      if constexpr (random_access_range<const _View> && sized_range<const _View>) {
```
This overload is non-const, so you should check the non-const _View

note that the requires clause uses const for a completely different reason: the spec tries to remove this non-const instantiation if non-const and const overload are exactly the same (we (sg9) always assume that  if the type of begin and begin const are the same , they are exactly the same). So the requires clause checks const to match the const overload. 

Here you are simply using the non-const base view so the check should be on non const View

casavaca replied:
You're right. Thank you.

huixie90 commented on libcxx/include/__ranges/drop_view.h:
```diff
@@ -90,6 +90,10 @@ namespace ranges {
     {
+      if constexpr (random_access_range<const _View> && sized_range<const _View>) {
+        auto __dist = ranges::distance(__base_);
+        return ranges::begin(__base_) + std::min<range_difference_t<_View>>(__count_, __dist);
```
I think you don’t need to explicit template parameter here because they are the same type.  If you want clarity, I would replace the ‘auto’ above with range_difference_t

casavaca replied:
Yep, got it.

huixie90 commented on libcxx/include/__ranges/drop_view.h:
```diff
@@ -90,6 +90,10 @@ namespace ranges {
     {
+      if constexpr (random_access_range<const _View> && sized_range<const _View>) {
+        auto __dist = ranges::distance(__base_);
+        return ranges::begin(__base_) + std::min<range_difference_t<_View>>(__count_, __dist);
```
Btw, please double check if this file has undef the min/max macro on the top

casavaca replied:
Not found anything similar. Besides, `std::min` is also used in other places of this file.

huixie90 replied:
yes there is already push/pop macros in this file

huixie90 commented on libcxx/test/std/ranges/range.adaptors/range.drop/begin.pass.cpp:
```diff
@@ -81,6 +81,36 @@ constexpr bool test() {
+    static_assert(std::ranges::sized_range<const NonCommonView>);
+    std::ranges::drop_view dropView9(NonCommonView{{}, 0, &sentinel_cmp_calls}, 4);
+    assert(dropView9.begin() == globalBuff + 4);
+    assert(sentinel_cmp_calls == 0);
```
The tests under libcxx/std are used by other implementations too (e.g. MSVC STL). It is not required by the spec that it has to be called zero times. I would instead relax this check a bit. For example , calling begin in the loop and check that we is not grow linearly ?

casavaca replied:
Yep. I can relax the test by comparing if `sentinel_cmp_calls` is a constant, but I have a question:
Since this is not required by the standard, how should I add test like this? I mean, should we use
something like `if defined(LIBCXX_ONLY_TESTS)`? or other other way? @huixie90

huixie90 replied:
For this one i think it would be good to keep it once you relaxed it as the constant time begin is required by the standard.  

In general if you have Libc++ specific test, you can put it under  test/libcxx

hewillk commented on libcxx/include/__ranges/drop_view.h:
```diff
@@ -90,6 +90,10 @@ namespace ranges {
                   random_access_range<const _View> && sized_range<const _View>))
     {
+      if constexpr (random_access_range<_View> && sized_range<_View>) {
+        range_difference_t<_View> __dist = ranges::distance(__base_);
```
```suggestion
        const auto __dist = ranges::distance(__base_);
```

huixie90 replied:
I was like why don't you use `ranges::size` here and that is exactly what you want to call since that is what is constrained on.  then I realised that oh you want to call `std::min` on two signed types so you used `ranges::distance` just for converting the types.  maybe 
```range_difference_t<_View> __dist = ranges::size(__base_);```
is the actual intent of the code?

hewillk replied:
> maybe
> `range_difference_t<_View> __dist = ranges::size(__base_);`
> is the actual intent of the code?

The standard does not guarantee that the above implicit conversion is well-formed. `ranges::distance` is always preferred over `ranges::size` when deal with iterators.

huixie90 replied:
> > maybe
> > `range_difference_t<_View> __dist = ranges::size(__base_);`
> > is the actual intent of the code?
> 
> The standard does not guarantee that the above implicit conversion is well-formed. `ranges::distance` is always preferred over `ranges::size` when deal with iterators.

Only narrowing conversion can be explicit-only.  IIUC if  size_t -> difeerence_t is narrowing conversion, I think we have other problems.  The problem of ranges::distance is that it is not for the purpose of calculating the size in constant time. It is a utility tool to calculate the size in case you cannot caluculate it in constant time, which is the wrong tool here imo. And I strongly against using ‘auto’ when dealing with difference type for the exact reason we are discussing: narrowing conversion. 
If the spec does not say size -> difference_t cannot be narrowing, I’d prefer to use either parenthesis or static_cast to do the comversion

hewillk replied:
> Only narrowing conversion can be explicit-only. IIUC if size_t -> difeerence_t is narrowing conversion, I think we have other problems. 

The standard does _not_ require the signedness of the return value of `ranges::size`. The only guarantee is that it is an integer-like type. So in theory, the types returned by `ranges::size` and `ranges::distance` can be unrelated and only be _explicitly_ converted.

> The problem of ranges::distance is that it is not for the purpose of calculating the size in constant time. 

This is why the `sized_range` needs to be satisfied in such case isn't it? in which case since `ranges::distance` is _guaranteed_ to be constant-time.

huixie90 replied:
> > Only narrowing conversion can be explicit-only. IIUC if size_t -> difeerence_t is narrowing conversion, I think we have other problems.
> 
> The standard does _not_ require the signedness of the return value of `ranges::size`. The only guarantee is that it is an integer-like type. So in theory, the types returned by `ranges::size` and `ranges::distance` can be unrelated and only be _explicitly_ converted.
> 
> > The problem of ranges::distance is that it is not for the purpose of calculating the size in constant time.
> 
> This is why the `sized_range` needs to be satisfied, isn't it? in which case `ranges::distance` is _guaranteed_ to be constant-time by its definition.

If you have a helper

Void helper() {
  if (c) 
       F();
   else
       G();
}

When you definitely don’t want to call G(), why do you write

if(c)
  helper()

Rather than call f() directly ?

huixie90 replied:
> > Only narrowing conversion can be explicit-only. IIUC if size_t -> difeerence_t is narrowing conversion, I think we have other problems.
> 
> The standard does _not_ require the signedness of the return value of `ranges::size`. The only guarantee is that it is an integer-like type. So in theory, the types returned by `ranges::size` and `ranges::distance` can be unrelated and only be _explicitly_ converted.
> 
> > The problem of ranges::distance is that it is not for the purpose of calculating the size in constant time.
> 
> This is why the `sized_range` needs to be satisfied, isn't it? in which case `ranges::distance` is _guaranteed_ to be constant-time by its definition.

Anyway I think our discussion about coding style should not block this patch to be landed

hewillk replied:
> If you have a helper
> 
> Void helper() { if (c) F(); else G(); }
> 
> When you definitely don’t want to call G(), why do you write
> 
> if(c) helper()
> 
> Rather than call f() directly ?

This example is not quite accurate because the return type of `ranges::distance` is what we want, which reduces the type conversion since the former has already been done for us. Using `ranges::size` requires an additional `static_cast`.

I believe the standard prefers using `rangs::begin(r) + rangs::distance(r)` rather than `rangs::begin(r) + static_cast<range_difference_t<R>>(rangs::size(r))`, given the former's heavy presence in the `<ranges>`'s wording.

huixie90 replied:
I am fine with both. But I think the spec is describing the behaviour and not necessarily the way one should code. e.g this particular function here the spec is simply describing what the result should be, “Returns” , and one should not use the exact wording for implementation . (I saw you create an LWG issue for this but lots of us think it is not a defect because the spec does not say “equivalent to”)

hewillk commented on libcxx/include/__ranges/drop_view.h:
```diff
@@ -104,7 +108,8 @@ namespace ranges {
       requires random_access_range<const _View> && sized_range<const _View>
     {
-      return ranges::next(ranges::begin(__base_), __count_, ranges::end(__base_));
+      range_difference_t<_View> __dist = ranges::distance(__base_);
```
```suggestion
      const auto __dist = ranges::distance(__base_);
```

hewillk reviewed (commented):
I don't think explicitly specifying the type improves readability, `ranges::distance` already implies signed values.

casavaca commented:
@huixie90 Is there anything I need to do? do you want me to change `range_difference_t<_View> __dist` to `const auto __dist`?

huixie90 commented on libcxx/include/__ranges/drop_view.h:
```diff
@@ -104,7 +108,8 @@ namespace ranges {
     {
-      return ranges::next(ranges::begin(__base_), __count_, ranges::end(__base_));
+      range_difference_t<_View> __dist = ranges::distance(__base_);
+      return ranges::begin(__base_) + std::min(__count_, __dist);
```
super nit:
the type of `(__count_)` is `range_difference_t<_View> const&` and the type of `(__dist)` is `range_difference_t<const _View> &`  and I am not sure `range_difference_t<_View>` and `range_difference_t<const _View>` have to be the same type.

hawkinsw replied:
For my own edification, I attempted to answer this question ... it looks like the type is stripped of `const`/`&`?

```C++
template <range _Rp>
using range_difference_t = iter_difference_t<iterator_t<_Rp>>;

// ...

template <class _Ip>
using iter_difference_t = typename conditional_t<__is_primary_template<iterator_traits<remove_cvref_t<_Ip> > >::value,
                                                 incrementable_traits<remove_cvref_t<_Ip> >,
                                                 iterator_traits<remove_cvref_t<_Ip> > >::difference_type;

```

huixie90 replied:
> For my own edification, I attempted to answer this question ... it looks like the type is stripped of `const`/`&`?
> 
> ```c++
> template <range _Rp>
> using range_difference_t = iter_difference_t<iterator_t<_Rp>>;
> 
> // ...
> 
> template <class _Ip>
> using iter_difference_t = typename conditional_t<__is_primary_template<iterator_traits<remove_cvref_t<_Ip> > >::value,
>                                                  incrementable_traits<remove_cvref_t<_Ip> >,
>                                                  iterator_traits<remove_cvref_t<_Ip> > >::difference_type;
> ```

I think this is unrelated. off course these traits needs to strip off `cvref`, but `iterator_t<_Rp>` and `iterator_t<const _Rp>` can well be two different types.  (although one would argue that it is evil to let them have two different difference_type)

hawkinsw replied:
Good catch! I missed the nested `iterator_t` inside the definition of the `range_difference_t`. 

And, you're right, that would be evil!

huixie90 reviewed (approved):
LGTM with green CI

casavaca commented:
Rebased on llvm/main. Also made stylish chanege to use `const auto`, similar to microsoft STL in 
https://github.com/microsoft/STL/blob/cf1313c39169dc376761eddee23c5e408e01aaa9/stl/inc/ranges#L3359

Let's see if CI is happy.

casavaca committed:
Update comment (remove zero width space)

huixie90 commented:
Hi, all CI passed now. could you please rebase and resolve the conflicts? It should be just a formatting change that had the conflicts. Thank you for your contribution

casavaca committed:
rebase
